common/src/androidTest/java/com/microsoft/identity/common/MsalCppOAuth2TokenCacheTest.java

```diff
@@ -0,0 +1,249 @@
+                LOCAL_ACCOUNT_ID,
+                USERNAME,
+                HOME_ACCOUNT_ID,
+                ENVIRONMENT,
+                REALM,
+                TARGET,
+                CACHED_AT,
+                EXPIRES_ON,
+                SECRET,
+                CLIENT_ID,
+                SECRET,
+                MOCK_ID_TOKEN_WITH_CLAIMS,
+                null,
+                CredentialType.V1IdToken
+        );
+    }
+
+    @After
+    public void tearDown() {
+        mCppCache.clearCache();
+    }
+
+    @Test
+    public void saveAndGetAccountTest() throws ClientException {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+
+        // Save the the cache
+        mCppCache.saveAccountRecord(generatedAccount);
+
+        // Restore it
+        final AccountRecord restoredAccount = mCppCache.getAccount(
+                generatedAccount.getHomeAccountId(),
+                generatedAccount.getEnvironment(),
+                generatedAccount.getRealm()
+        );
+
+        Assert.assertNotNull(restoredAccount);
+        Assert.assertEquals(generatedAccount, restoredAccount);
+    }
+
+    @Test
+    public void getAccountNullTest() throws ClientException {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+        final AccountRecord restoredAccount = mCppCache.getAccount(
+                generatedAccount.getHomeAccountId(),
+                generatedAccount.getEnvironment(),
+                generatedAccount.getRealm()
+        );
+        Assert.assertNull(restoredAccount);
+    }
+
+    @Test
+    public void getAllAccountsTest() {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+        mCppCache.saveAccountRecord(generatedAccount);
+
+        final List<AccountRecord> accounts = mCppCache.getAllAccounts();
+        final AccountRecord restoredAccount = accounts.get(0);
+        Assert.assertEquals(generatedAccount, restoredAccount);
+
+        // TODO this API makes no statement about the immutability of the List returned
+        // TODO results should not be mutable
+    }
+
+    @Test
+    public void getAllAccountsEmptyTest() {
+        final List<AccountRecord> accounts = mCppCache.getAllAccounts();
+        Assert.assertTrue(accounts.isEmpty());
+    }
+
+    @Ignore // Ignore this test until API behavior is decided
+    @Test
+    public void removeAccountTest() throws ClientException {
+        final AccountRecord generatedAccount = mTestBundle.mGeneratedAccount;
+        mCppCache.saveAccountRecord(generatedAccount);
+        final AccountDeletionRecord deletionRecord = mCppCache.removeAccount(
+                generatedAccount.getHomeAccountId(),
+                generatedAccount.getEnvironment(),
+                generatedAccount.getRealm()
+        );
+        Assert.assertEquals(generatedAccount, deletionRecord.get(0));
+
+        // TODO There's confusing behavior here...
```

Comment:
👀

Reply:
What's the use case for removeAccount?  It seems like we either need to remove it from *all* clients (because, say, this is called when the user is being removed from their home tenant) or we need to be able to specify the client we're removing the account from (because, say, the user is being removed from an application).  Either way, the use case should tell you which way to go.